File: typescript/Models/SolutionModels.ts
const SOLUTION_CONFIG_KEY = "Solution" as const;

enum SolutionType {
  First = "First",
  Second = "Second",
}

enum RunType {
  Test = "Test",
  Full = "Full",
}

type SolutionOptionsInit = {
  day?: number | string;
  Day?: number | string;
  solutionType?: SolutionType | keyof typeof SolutionType | string;
  SolutionType?: SolutionType | keyof typeof SolutionType | string;
  runType?: RunType | keyof typeof RunType | string;
  RunType?: RunType | keyof typeof RunType | string;
};

class SolutionOptions {
  public day = 1;
  public solutionType: SolutionType = SolutionType.First;
  public runType: RunType = RunType.Test;

  public constructor(init?: SolutionOptionsInit) {
    if (init) {
      this.apply(init);
    }
  }

  public apply(init: SolutionOptionsInit): void {
    const desiredDay = init.day ?? init.Day;
    if (desiredDay !== undefined) {
      const parsed = Number(desiredDay);
      this.day = Number.isNaN(parsed) ? this.day : parsed;
    }

    const desiredSolutionType = init.solutionType ?? init.SolutionType;
    if (desiredSolutionType !== undefined) {
      this.solutionType = this.parseSolutionType(desiredSolutionType);
    }

    const desiredRunType = init.runType ?? init.RunType;
    if (desiredRunType !== undefined) {
      this.runType = this.parseRunType(desiredRunType);
    }
  }

  private parseSolutionType(value: SolutionOptionsInit["solutionType"]): SolutionType {
    if (typeof value === "string") {
      const normalized = value.trim();
      if (normalized in SolutionType) {
        return SolutionType[normalized as keyof typeof SolutionType];
      }

      const match = (Object.values(SolutionType) as string[]).find(
        (entry) => entry.toLowerCase() === normalized.toLowerCase(),
      );

      if (match) {
        return match as SolutionType;
      }
    } else if (typeof value === "number") {
      const enumValues = Object.values(SolutionType);
      return (enumValues[value] as SolutionType) ?? this.solutionType;
    } else if (value) {
      return value;
    }

    return this.solutionType;
  }

  private parseRunType(value: SolutionOptionsInit["runType"]): RunType {
    if (typeof value === "string") {
      const normalized = value.trim();
      if (normalized in RunType) {
        return RunType[normalized as keyof typeof RunType];
      }

      const match = (Object.values(RunType) as string[]).find(
        (entry) => entry.toLowerCase() === normalized.toLowerCase(),
      );

      if (match) {
        return match as RunType;
      }
    } else if (typeof value === "number") {
      const enumValues = Object.values(RunType);
      return (enumValues[value] as RunType) ?? this.runType;
    } else if (value) {
      return value;
    }

    return this.runType;
  }
}

export { RunType, SolutionOptions, SolutionType, SOLUTION_CONFIG_KEY };

File: python/models/solution_models.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any, Mapping

SOLUTION_CONFIG_KEY = "Solution"


class SolutionType(str, Enum):
    FIRST = "First"
    SECOND = "Second"


class RunType(str, Enum):
    TEST = "Test"
    FULL = "Full"


@dataclass
class SolutionOptions:
    day: int = 1
    solution_type: SolutionType = SolutionType.FIRST
    run_type: RunType = RunType.TEST

    @staticmethod
    def from_dict(data: Mapping[str, Any]) -> "SolutionOptions":
        options = SolutionOptions()
        options.apply(data)
        return options

    def apply(self, data: Mapping[str, Any]) -> None:
        if "Day" in data or "day" in data:
            day_value = data.get("Day", data.get("day"))
            if isinstance(day_value, str):
                day_value = day_value.strip()
                if day_value.isdigit():
                    self.day = int(day_value)
            elif isinstance(day_value, (int, float)):
                self.day = int(day_value)

        if "SolutionType" in data or "solutionType" in data:
            raw_value = data.get("SolutionType", data.get("solutionType"))
            parsed = self._parse_enum(raw_value, SolutionType, self.solution_type)
            self.solution_type = parsed

        if "RunType" in data or "runType" in data:
            raw_value = data.get("RunType", data.get("runType"))
            parsed = self._parse_enum(raw_value, RunType, self.run_type)
            self.run_type = parsed

    @staticmethod
    def _parse_enum(value: Any, enum_type: type[Enum], fallback: Enum) -> Enum:
        if isinstance(value, enum_type):
            return value
        if isinstance(value, str):
            normalized = value.strip()
            for member in enum_type:
                if member.name.lower() == normalized.lower():
                    return member
                if member.value.lower() == normalized.lower():
                    return member
        return fallback


__all__ = ["SolutionOptions", "SolutionType", "RunType", "SOLUTION_CONFIG_KEY"]

